Refactor EIP712 to be consistent with MetaMask

thomas-waite opened the pull request:
## Summary
This PR makes the AZTEC EIP712 implementation consistent with that of MetaMask. 

MetaMask uses the package `eth-sig-util` to perform signing, specifically using the function `signTypedData()` - https://github.com/MetaMask/eth-sig-util. This outputs an ECDSA signature, with the parameters being output in the order: `(r, s, v)`.

This order of parameters is different to the AZTEC signing functions, which output in order`(v, r, s)`. The EVM precompile that is then used to perform signature verification, ecrecover, expects params to be in order `(v, r, s)`. 

This PR updates the AZTEC signature verification contracts and signing functions, so that signatures generated with MetaMask (signing over AZTEC notes) can be verified by the AZTEC contracts.

<!--- Describe your changes in detail -->
## Detailed changes
The major changes relate to two packages: `aztec.js` and `protocol`.

**aztec.js**
- `signer` module: Where the note signing functions return ECDSA parameters, they have been adjusted to be in the order `(r, s, v)` rather than `(v, r, s)`
- `proof` module: the `joinSplit.constructSignatures()` code has been refactored and largely broken out into a seperate method that is now in the `signer` module

**protocol**
- `LibEIP712.sol` - `recoverSignature()` adapted to deal with params in `(r, s, v)` order and rearrange into correct form for ecrecover precompile
- `ZkAsset.sol` - `extractSignature()`adapted to deal with params in different order

## Testing instructions
`yarn test`

Various additional tests have been added

## Types of changes
**Breaking change**
There are a couple of breaking changes:
1) `signer.signNote()` has been renamed to `signer.signNoteForConfidentialApprove()` to better describe what this function should be used for and seperate it from the other signing functions
2) The ECDSA parameter output order for note signing functions in the `signer` module have been changed from `(v, r, s)` to `(r, s, v)`

thomas-waite committed:
refactor(protocol): make sig verify contract methods match MetaMask
MetaMask signTypedData() function outputs ecdsa params in order: (r, s, v)
adjusting aztec signing functions to match this
ecrecover precompile expects ecdsa params to be in order: (v, r, s)
so LibEIP712.recoverSignature() and ZkAsset.extractSignature() adjusted to arrange ecdsa params into order required for precompile

thomas-waite committed:
refactor(aztec.js): make sig ecdsa param order match MetaMask
MetaMask signTypedData() outputs as (r, s, v)
change signer.signNote() to match this
BREAKING CHANGE: change sig ecdsa param order for signer.signNote()

thomas-waite committed:
test: add tests to check sign functions
check publicKey can be recovered, v is well formed
check AZTEC signing function output matches eth-sig-util.signTypedData() signing function output

thomas-waite committed:
refactor(aztec.js): create signNotesForConfidentialTransfer() function
break out the joinSplit proof constructSignatures() code into a seperate function in signer module
name it signNotesForConfidentialTransfer() to indicate the context in which this note signing function is used
rename signer.signNote() method to signer.signNoteForConfidentialApprove() to better indicate the context in which this signing function should be used
update docs, to be more clear about intended use case of signing functions
BREAKING CHANGE: rename signer.signNote() to signer.signNoteForConfidentialApprove()

thomas-waite committed:
docs(protocol): update docs and eslint fixes

thomas-waite committed:
test(protocol): update to renamed methods

joeandrews commented on packages/aztec.js/src/signer/index.js:
```diff
@@ -62,7 +64,43 @@ signer.signNote = (verifyingContract, noteHash, spender, privateKey) => {
+    const r = signature[1];
+    const s = signature[2].slice(2);
+    const v = signature[0].slice(-2);
+    return r + s + v;
```
Lets just have one function that returns the signature structure that is compatible with metamask (r + s + v).

joeandrews replied:
Ignore previous comments

thomas-waite replied:
Good idea, done

thomas-waite committed:
refactor(aztec.js): make signTypedData() output unformatted params
signTypedData() outputs unformatted params
formatting moved into higher level, wrapper signing functions
doc updates
BREAKING CHANGE: signTypedData() now outputs r, s, v as unformatted

thomas-waite committed:
test: migrate tests to updated signing methods

thomas-waite committed:
refactor: create signMultipleNotesForConfidentialTransfer() function

ArnSch commented:
@thomas-waite Did you notice the failure during tests?

thomas-waite commented:
yeah, fixing it today. Some problem that came about as a result of rebasing

thomas-waite committed:
test(protocol): fix LibEIP712.sol tests

ArnSch reviewed (approved):
LGTM
